### ISSUE TITLE
Add Neo2 preset keymap

### ISSUE THREAD

**Issue by davidak** [UltimateHackingKeyboard/agent#678]
Here is a new issue to track progress since https://github.com/UltimateHackingKeyboard/agent/issues/91 was closed without adding it.

You find the layout with all 6 layers here: https://www.neo-layout.org/

**Comment by mondalaci** [comment]
Thanks for the report! This is currently blocked by #617.

**Comment by tuxlifan** [comment]
In the meantime I created a "Reverse" map that can be used to get back the letters a-z like in QWERTY when the OS-keymap is set to Neo2.
i.e. when you switch to Neo2 on OS level

**Comment by bjoernjacobs** [comment]
Having Neo2 would be great. I just ordered my UHK and - to be honest - expected it to not have problems with any layout. Finding out that it is not supported yet (after I ordered) was a little bummer.

**Comment by mondalaci** [comment]
@bjoernjacobs You can already create a Neo2 keymap on the UHK. In case 4 layers aren't sufficient for you, I think [Karel's UHK firmware fork](https://github.com/kareltucek/firmware) provides more. We'll add more layers to the native UHK firmware and Agent eventually.

**Comment by bjoernjacobs** [comment]
@mondalaci Thank you for the fast feedback. If I can use the first 4 layers, I'm happy! (Layer 5 and 6 of Neo2 contain mainly greek letters which I do not need for coding)

**Comment by mondalaci** [comment]
If you guys think that Neo2 can already be implemented reasonably well on the UHK without #617 then please come up with a Neo2 UHK keymap, attach your configuration, and I'll add it to the official UHK configuration.

**Comment by qux42** [comment]
I build up neo2 layout on a English base layout. It works really good. Sadly I'm in holidays, so I can't export it at the moment.
I will do it at soon I'm back

Sam
________________________________

**Comment by skaldesh** [comment]
@qux42 Hi, I did the same as you. Did you get it to work so that after pressing both Mod3 keys, you enter "Capslock" mode? I did not manage to get it work (but I tried it 2 years ago, dont know what changed)

**Comment by evs-ch** [comment]
Just in case someone might find it useful to have something to start with, here's my config with a rather basic neo2 layout I've made ages ago when I got the UHK.
It works to have at least a basic neo-ish layout "in dire emergencies"; normally I just use the neo2 implementation in userspace with just a slightly modified qwerty layout on the UHK. No special unicode chars etc., of course, and whatever improvements there might've been in the recent month that might have allowed to implement it better are obviously not part of it yet :)

https://gist.github.com/evs-ch/4aa744a24c7a74e51b34f000d7299841

**Comment by skaldesh** [comment]
I created one as well and it is what I use every day. I am using macros to write ä,ö,ü,ß,…
The only thing not working for me (apart from missing layers 5 and 6, obviously) is I can not switch to Capslock mode and on layer 4 I only added the keys I normally need, so it is not complete

https://gist.github.com/skaldesh/6322b54e180c54691357bea0604c9922

**Comment by djmj0** [comment]
Hi,

i just received my UHK and i am totally impressed.

However, i am also an avid Neo2 user and as such i am looking forward to an official implementation to UHK.

Unfortunately, when trying to import the scripts from either @evs-ch or @skaldesh, i get the following error message: "Invalid configuration specified". Could someone point me in the right direction of what i am doing wrong? I downloaded both scripts as .json and tried to import them directly into Agent, please see the attached screenshot.

Thank you!

![Screenshot from 2020-05-20 18-07-34](https://user-images.githubusercontent.com/56806415/82469844-1a0be780-9ab4-11ea-9d4c-eb03019d52a3.png)

**Comment by mondalaci** [comment]
@djmj0 Thanks for the nice words and for reporting this issue!

@ert78gb I think Agent should be able import these configurations. Please look into this issue.

**Comment by djmj0** [comment]
thank you for your quick response, @mondalaci . 
I am happy to help if there are questions.

**Comment by lukaskawerau** [comment]
@mondalaci - just adding a +1 on the import problem, I can't import the linked config either.

**Comment by ert78gb** [comment]
Sorry for the late answer but I was very busy in the last few days.
I just now started to investigate the https://gist.github.com/skaldesh/6322b54e180c54691357bea0604c9922 file and there is 2 things:
- There is in the bug in the import. PR opening soon.
- The configuration is invalid. It contains "Switch Keymap" actions to keymaps that does not exists in the configuration like
  - DVO
  - COL
  - QWM
  - DVM
  - COM

My assumption is
- Configuration exported
- Manually deleted the unwanted shared keymaps.

**Comment by ert78gb** [comment]
The https://gist.github.com/evs-ch/4aa744a24c7a74e51b34f000d7299841 file has a same problem.
Missing keymaps:
- ADM
- QWR
- QWM
- COM
-

**Comment by mondalaci** [comment]
@ert78gb Regarding the invalid configuration file, please manually remove invalid switch keymap actions and reattach the file.

**Comment by ert78gb** [comment]
@mondalaci Ok, but in this case, these configs will contain only the neo2 keymaps.

**Comment by mondalaci** [comment]
@ert78gb Then please use the default configuration and add the Neo2 keymap. Both of the above configs are affected? If so, please fix both.

**Comment by ert78gb** [comment]
Both affected.
The default user config extended with Neo2 config(s):
Need to unzip them before the import
[default-user-configuration-with-skaldesh-neo2.json.zip](https://github.com/UltimateHackingKeyboard/agent/files/4672429/default-user-configuration-with-skaldesh-neo2.json.zip)
[default-user-configuration-with-evs-ch-neo2.json.zip](https://github.com/UltimateHackingKeyboard/agent/files/4672431/default-user-configuration-with-evs-ch-neo2.json.zip)

**Comment by ert78gb** [comment]
The user config that contains all the 3 Neo2 config
[default-user-configuration-with-3-neo2.json.zip](https://github.com/UltimateHackingKeyboard/agent/files/4672516/default-user-configuration-with-3-neo2.json.zip)

**Comment by mondalaci** [comment]
As it turned out, the import config bug which has just been fixed by #1254 only affected the web demo of Agent. It seems that the attached configurations were edited by hand and are invalid.

@davidak @2er0 @tuxlifan @bjoernjacobs @qux42 @skaldesh @evs-ch @djmj0 @lukaskawerau Please export your current configuration for backup purposes, then download, unzip, and import the [default-user-configuration-with-3-neo2.json.zip](https://github.com/UltimateHackingKeyboard/agent/files/4672530/default-user-configuration-with-3-neo2.json.zip) configuration.

This configuration contains the `evs-ch forneo`, `evs-ch semineo`, and `skaldesh Neo` keymaps. I'd like you to reach a consensus on which one of these is the best Neo2 keymap, and then we'll add it to Agent as a preset.

Please note that the related macros contained by @skaldesh's original keymap were removed as no macros can be used for keymap presets.

**Comment by lukaskawerau** [comment]
Thank you @ert78gb and @mondalaci for taking this on!  
I'm very, very, very new to Neo, so my opinion should probably count the least when discussing which preset to use, but as far as I can tell @skaldesh's keymap seems to be the closest to the full Neo2 layout. I.e. in the lower left I think I see where the Umlaut-Macros were.

Since I'm using UHK on the Mac, will there be a Mac and PC version, like there is for Dvorak and Colemak?

**Comment by mondalaci** [comment]
I'm fine with having both Mac and PC Neo2 preset keymaps as long as you guys create them, and agree that they're the real deal.

**Comment by skaldesh** [comment]
Hey guys,  
I would love to agree on one layout. Before we use mine, for example, as a basis, I would like to ask some questions:

> Please note that the related macros contained by @skaldesh's original keymap were removed as no macros can be used for keymap presets.  

@mondalaci is there any way to export macros as well? Without macros, some keys for Neo are not possible to write with the current UHK agent (mainly "…", "ä", "ö", "ü"). Maybe the macros could be offered in an extra file, for people who want to get as close to the Neo2 keymap as possible.

Furthermore, a huge disadvantage with my layout right now is that you can not press both Shift keys to switch to "Capslock" (in Neo2, the second layer). This is one of the greatest things in Neo2, that you are able to type in Capslock, but can still type all other special characters. I have not checked up on the features of the UHK in quite a while, so I am not sure whether something like this can be achieved now. It was not possible back when I created my keymap, though.  

In case we want to use my keymap as basis (or any other), I can only help with PC Neo2 preset keymaps. @lukaskawerau you can do the Mac version, right?

**Comment by lukaskawerau** [comment]
This is the most faithful representation I could come up with, based on the Neo layout presented [here](https://git.neo-layout.org/neo/neo-layout/wiki/Grafiken).  As @skaldesh says, the biggest problem atm is that the second NEO2 layer (Shift) isn't programmable, and some of the keys Neo2 requires aren't typable in macOS with a single stroke: on the US keyboard, ü requires Option+u+u to type `¨` first and then place a `u` under it.  

<img width="1010" alt="Screenshot 2020-05-27 at 15 34 36" src="https://user-images.githubusercontent.com/2502467/83026929-ba6f8800-a02f-11ea-97d1-1b4dd4af03fa.png">
<img width="1010" alt="Screenshot 2020-05-27 at 15 34 41" src="https://user-images.githubusercontent.com/2502467/83026935-bd6a7880-a02f-11ea-8bea-bd5f51150cdc.png">
<img width="1010" alt="Screenshot 2020-05-27 at 15 34 45" src="https://user-images.githubusercontent.com/2502467/83026937-be030f00-a02f-11ea-8f43-d9473837e438.png">
<img width="1010" alt="Screenshot 2020-05-27 at 15 34 48" src="https://user-images.githubusercontent.com/2502467/83026941-be9ba580-a02f-11ea-9bfa-4b10167fe350.png">

**Comment by mondalaci** [comment]
@skaldesh The macros were removed to make preset keymaps self-contained and less complicated conceptually. It'd be quite strange from a user standpoint if macros were added when adding preset keymaps.

Do you use the US OS layout? I'm asking because I think by using DE, the "ä", "ö", "ü" characters would be easy to trigger without macros. I'm wondering whether every preset keymap should be created (even the ones besides Neo2) assuming that the user uses the US OS layout.

Layer switching via both Shift keys might be possible via [Karel's UHK firmware fork](https://github.com/kareltucek/firmware), but I don' want to merge such niche features because, again, they'd make Agent less conceptually clean.

I'd much rather prefer using built-in UHK features for layer switching. Using the secondary role of Tab might be feasible for example.

@lukaskawerau Thanks for the great work! I'm trying to wrap my head around Neo2, but I'm unsure about how should its layers be triggered. There's a limited amount of English resources available, and I only found [this Wikipedia section](https://en.wikipedia.org/wiki/Neo_(keyboard_layout)#Layers) regarding the layers. According to this article, is it also possible to switch layers with "#" and "<" keys? Oddly, the both Shift layer switching method mentioned by @skaldesh is not even mentioned here. Some explanation on this would be much appreciated.

**Comment by lukaskawerau** [comment]
@mondalaci Happy to help!

One additional challenge is how to treat ANSI and ISO layouts - the way I've set up the layers Mod corresponds to Mod3 in Neo (-> Layer 3) and Fn corresponds to Mod4 (-> Layer 4) in Neo. Because ANSI and ISO differ, so does the placement of layer switchers.  
Since I'm new to Neo2 as well, @skaldesh will need to comment on locking Layer 2 by tapping both shift keys.  
But the general point is: since UHK doesn't expose the "Shift-Layer" (Neo Layer 2) for editing, that means that symbols such as € can't be placed where they would in Neo2.

**Comment by kareltucek** [comment]
(Layer switching with both shifts is possible on the fork (e.g., via `ifKeyActive` and `toggleLayer`), but the macro layer mechanism is not integrated with the official layer mechanism, which means that one may need  to either to be careful about untoggling the layer again via some macro, or to entirelly switch to the macro induced mechanism, or to fire a ticket with description of the encountered problems and keep kicking me until I implement some fix.) 

(More than four layers are also possible as mentioned above, but face the same limitation.)

**Comment by skaldesh** [comment]
@kareltucek  Wow, I am just browsing your fork and it looks amazing! I need to spent some time with it, maybe I can finally have a fully functioning Neo2 layout.  

@mondalaci Yes, I am using US layout.  

Since I can't do anything with the regular features of the UHK to complete the Neo2 keymap, my layout config I posted above can be considered the best I can do. I would argue against making it some kind of ofiicial Neo2 preset, since it simply is not real Neo2.

**Comment by mondalaci** [comment]
@skaldesh When #617 gets done, will you be able to implement a UHK Neo2 keymap that you'd consider complete?

Is double Shift locking an official feature of Neo2?

**Comment by skaldesh** [comment]
> Is double Shift locking an official feature of Neo2?  

I think we overemphasized this a little bit. It is nothing more than the QUERTY equivalent of "Capslock". In Neo2, the Capslock key is used to activate Mod3 layer. By pressing both shift keys, you can permanently activate Mod2 layer (uppercase letters). The "special thing" is that whilst I have activated the Capslock mode, I can still press for example the Capslock key to activate Mod3 layer. This makes it possible to type in uppercase and still access all puncuation, numbers, arrow keys, etc etc like you would normally.  
I hope I could clear it up a little bit.  

Regarding whether this is an official feature, yes it is:  
> Drückt man die beiden Umschalttasten gleichzeitig, wird CAPSLOCK aktiviert. Und genauso deaktiviert.  
> -- <cite>https://git.neo-layout.org/neo/neo-layout/wiki/Tastaturbelegung#Ebene2</cite>

My rough translation: By pressing both Shift keys at the same time, CAPSLOCK is activated. And also deactivated [by pressing again]  

There is also something mentioned on Wikipedia in section **Layers** of the Neo2 article:
> The fourth level can be reached via Mod4 , which under QWERTY corresponds to the Alt Gr key and the < key, contains a numeric keypad and important navigation keys, so you do not have to take your hands off the main field to navigate in a text document. By making the navigation buttons accessible on the main panel, Neo also encounters the criticism expressed in reform keyboards that text editors like Vim would be more difficult to use. **This level can be locked just like the second one**.  
> -- https://en.wikipedia.org/wiki/Neo_(keyboard_layout)

**Comment by skaldesh** [comment]
> @skaldesh When #617 gets done, will you be able to implement a UHK Neo2 keymap that you'd consider complete?  

Tbh, I am not sure what #617 actually allows me to do. More layers sound promising, but what I miss the most for Neo2 is still the Neo2 capslock mode as described above, where I can access a layer (Mod3 and Mod4) from another layer (Mod2 in this case). Is that possible with the issue?

**Comment by mondalaci** [comment]
@skaldesh Thanks for the explanation! #617 may be useful for manipulating the scancodes of keys while modifiers are used, and more layers might also be useful, but it doesn't allow for the double shift feature of Neo2. We might merge some of Karel's advanced features over time to allow for double shift, but I don't plan to make a dedicated UI for such niche features in Agent.

**Comment by skaldesh** [comment]
> We might merge some of Karel's advanced features over time to allow for double shift, but I don't plan to make a dedicated UI for such niche features in Agent.   
!!! This is the way to go !!!  

I understand your desire to keep the agent simple, there are countless discussions in this repo's issue about it. If we can make the "expert" features accessible in another way, I am totally fine with it.

**Comment by kareltucek** [comment]
> We might merge some of Karel's advanced features over time to allow for double shift, but I don't plan to make a dedicated UI for such niche features in Agent.

(It surely sounds good, but has a fair number of catches. Mainly, the command-driven way of triggering events also depends on the strongly generalized macro engine of the fork (since to be of any use, it has to be able to run multiple macros parallely along standard operation of the keyboard). While unification of the two implementations is surely desirable from my side, I cannot deny that the implementation is in a pretty bad shape and porting it partially or refactoring would be difficult.)

**Comment by skaldesh** [comment]
Update:  
I discovered today, that I can simply activate the Capslock mode of the UHK to simulate the double  Shift lock.  
In the agent, I mapped an unused key in my Mod layer to activate Capslock.    

NOW_I_CAN_TYPE_LIKE_THIS very comfortably...

**Comment by mondalaci** [comment]
@skaldesh So glad to hear! Is this a suitable replacement of the double Shift lock, and does it work on all OSes?

**Comment by skaldesh** [comment]
It should work on all OSes, though, I can confirm it only for Linux right now.  
For me, I can live with it. Still, I would not be too keen on labeling it as a *Neo2 Preset* :)

**Comment by mondalaci** [comment]
@skaldesh Gotcha, thanks! If I understand correctly, only double shift lock is missing for a pristine Neo2 UHK experience, but it can be simulated reasonable well via Caps Lock. Please feel free to attach your updated Neo 2 keymap, and we'll include it eventually into Agent.

**Comment by skaldesh** [comment]
> If I understand correctly, only double shift lock is missing for a pristine Neo2 UHK experience  

For me, yes. But I am just a single person, others might say that they miss more keys/features from Neo2 (my Mod3 and Mod4 layer still lack a lot of keys of Neo2, since I am either not using them or sacrificed them to move the mouse).  
Also, the thing about macros not being included in exported keymaps leads to the problem that the german Umlaute (ä, ö, ü) and the three dots "…" are missing in my keymap. Again, might be ok for some people but critical for others (I would consider myself as part of the latter). 

> Please feel free to attach your updated Neo 2 keymap  

I will try one more time to polish some things, then upload it here.

**Comment by skaldesh** [comment]
@mondalaci I would export my keymap now again. Should I just export the whole device configuration via the Agent and upload it here?

**Comment by mondalaci** [comment]
@skaldesh Yes, that'd work. Just make sure not to upload sensitive macros.

**Comment by skaldesh** [comment]
[UserConfiguration.json](https://github.com/UltimateHackingKeyboard/agent/files/4769643/UserConfiguration.txt)

**Comment by Akku-21** [comment]
For me the german Umlaute work, if I map certain punctuations on the key.
ü={[
ö=:;
ä="'

and the macros from skaldesh dont work for me.

**Comment by skaldesh** [comment]
@Akku-21  
You are probably using a gerrman keyboard layout on your OS, that is why you can map the punctuation characters to the Umlaute. My configuration is targeted to US keyboard layouts, as it is the most common I have to deal with on other machines (servers, embedded systems, etc.).  
My macros are not exported in the last configuration and seem to have been dysfunctional in my first export attempt. If you want them, I can try to give them to you

**Comment by poelzi** [comment]
I got my uhk on friday and started on getting neo on the board. The keyboard is awesome, but the agent, especially in respect to mapping non us based layouts to uhk could definitely need some more work. I will give  #617 a try, since the default layer configuration is definitly not enough. Even that the shift layer is linked to the base layer is problematic.
The umlaut macros don't work for me.
Could it be, that neo can only be mapped correctly if or a special keymap is used on the pc, since there are many none us based characters. Thinking about layer 5 and 6.

**Comment by skaldesh** [comment]
> The umlaut macros don't work for me.
  Could it be, that neo can only be mapped correctly if or a special keymap is used on the pc, since there are many none us based characters. Thinking about layer 5 and 6.

I think UHK now supports arbitrary scancodes with #857. Have not tested it myself. Please report back, if you get it to work as expected :)

**Comment by mondalaci** [comment]
@poelzi #617 is not ready yet. [Karel's UHK firmware fork](https://github.com/kareltucek/firmware) enables advanced use cases you may be looking for. As for the umlauts, please read [this article](https://ultimatehackingkeyboard.com/blog/2018/06/23/how-can-i-type-accented-characters-with-my-uhk).

**Comment by mondalaci** [comment]
Now that Agent supports 12 layers, would some of you submit a Neo2 keymap, guys? We'd gladly add it to keymap presets.

**Comment by skaldesh** [comment]
I can do it :)  
I will have to tinker around a little bit with the new agent first. Haven't used it in quite a while because my UHK does what I need it to do.

**Comment by mhantsch** [comment]
I'm trying to create one, based on a German base PC keymap. I think I too need to learn some features of the new agent and firmware now. Like, how do I create a non-shifted keypress on the shift layer? It seems like the shift mod always gets added to the keypress. Do I need to revert to macros for this?

**Comment by kareltucek** [comment]
> Like, how do I create a non-shifted keypress on the shift layer? It seems like the shift mod always gets added to the keypress. Do I need to revert to macros for this?

You just map a non-shift keypress in the shift layer, save the config and then tap the key with regular shift modifier (not, e.g., a shift macro).

If it does not work, please, create a new ticket with description of your setup.

**Comment by mhantsch** [comment]
> You just map a non-shift keypress in the shift layer

You are of course right; it works as advertised. I got myself confused with too many remappings... (editing the NEO map while using a Coleman keyboard map with a German keymap active in Windows).

I assume I will need macros though to access many of the special characters in Neo that can only be produced with some U+abcd codes.

**Comment by skaldesh** [comment]
> I assume I will need macros though to access many of the special characters in Neo that can only be produced with some U+abcd codes.  

If that is still the case, then Layer 5 + 6 will be a nightmare.  
Is #857 no help in this regard?

**Comment by mondalaci** [comment]
Please let's not use Unicode character sequences as they're OS-specific.

Also, we don't have a mechanism to add macros to preset keymaps, and they'd be confusing when adding new preset keymaps, so macros shouldn't be used either.

**Comment by mhantsch** [comment]
Unfortunately, I don't see a way to produce some of the Neo characters without the use of macros. Some of the characters just don't have a 1:1 mapping but require a sequence of keystrokes (very dependent on OS and chosen base OS keymap).

**Comment by skaldesh** [comment]
Probably the only way to have a real Neo2 keyboard layout on the UHK is to have Neo2 also configured on the OS side. Is that correct @mondalaci ?

**Comment by mondalaci** [comment]
I'm not sufficiently familiar with Neo2, but based on the above, it's not achievable without OS-specific macros. I think this layout is too problematic to be implemented in a general manner on any keyboard, so I'm closing this issue. Thanks for your input, guys!

**Comment by mhantsch** [comment]
@mondalaci Not Neo-specific: let's say I create this amazing layout which works together with a few macros (and needs them), and I want to share it with my friends (or the world). How do I export it into a format that someone else can import and use? Once you reference macros from your keymap, I think they belong with it.

**Comment by mondalaci** [comment]
I can see some problems with bundling macros together with keymaps, but I can also see the merit of the feature. However, this feature is only relevant for Neo2, which uses OS-specific macros, so I don't think it's worth pursuing.

**Comment by SirVer** [comment]
I've read through this very long thread and I wonder what the official statement now is. I've seen a few configurations posted, but they've seen quite old at this point - are they still the status quo for Neo2 support? I am basically using the same that I use for the build in keyboard in my macbook, i.e. a very karabiner elements based software solution. Would be great to have this more native with the UHK, but I also understand the maintainers, that Neo2 is a lot of work for probably only a handful of people.

so my question: what are people using for Neo2 on the UHK currently?

**Comment by 2er0** [comment]
I have the UHK with a QUERTY keymap and the standard Neo2 setup for MacOS (Keyboard layout, Karabiner). Plus, some special configurations in the UHK to have additional functionality.

It is possible to program Neo2 as a layout on the UHK with a Unicode Hex Input (Mac keyboard layout), but the UHK does not support the UTF-8 codes that Neo2 uses, which means only the first and second layer of Neo2 work without the üöä symbols. 

I use the fn and the key below the spacebar on the UHK for special modifications. Everything else is a standard Neo2 setup.

**Comment by mondalaci** [comment]
> but the UHK does not support the UTF-8 codes that Neo2 uses, which means only the first and second layer of Neo2 work without the üöä symbols.

What do you mean exactly? USB itself doesn't support Unicode natively, only scancodes. But OS-specific Unicode characters can be replicated via UHK macros.

**Comment by 2er0** [comment]
I could not program, for example, `ß`, `[` or others from layer 3 and 4 in Neo in combination with the Unicode Hex Input in MacOS. 
The current solution uses a QUERTY layout on the UHK with the normal Neo setup on MacOS. The key below the spacebar is for additional functionality, such as music control.

**Comment by 2er0** [comment]
Yes I did

**Comment by mondalaci** [comment]
Feel free to paste the screenshot of your macro. We'd look into it.
